File: panel/awg/names.py
"""Random names, for the things that are created without one.

A client and an API token are both named by whoever creates them, and in both
cases the name is an identifier rather than a label: a client's name is what its
config file is called and what every per-client route is addressed by, and a
token's name is what the activity log writes beside everything that token does.
So something has to be in the box, and an operator adding the fortieth client of
an afternoon should not have to invent the fortieth word.

Nine characters out of the alphabet below, drawn here so that both callers draw
them the same way.

The alphabet is lower case and leaves out `l`, `1`, `o` and `0`. These names are
read off a screen and typed back into a shell, a support message or another
panel, and those four are the pairs that get read wrong when they are; dropping
them costs a quarter of a bit per character and removes the one failure mode a
random name has. What is left satisfies both name rules the panel has - the
client name pattern in awg.store and the printable-text rule for a token - so a
name from here is never refused by the thing it was drawn for.

Thirty-two characters over nine positions is forty-five bits, which is why
`unique_name` exists and why it is not a loop anybody should expect to go round
twice. A collision is not the reason for the check: the reason is that the name
is an identifier, and an identifier that is merely very unlikely to be taken is
still one that can be, on the one server where it happens, silently overwriting
somebody's client if nothing looks first.
"""

import secrets
from collections.abc import Callable

# Deliberately not `random`: nothing here is a secret, but the process-wide
# generator can be seeded by anything that imports it, and names repeating after
# a restart is exactly what this module exists to avoid.
ALPHABET = "abcdefghijkmnpqrstuvwxyz23456789"

LENGTH = 9

# How many draws before giving up and saying so. Reaching the end of this means
# the caller's "is it taken" answered yes thirty-two times running, which on a
# space this size is not a full table - it is something wrong with the question.
ATTEMPTS = 32


def random_name(length: int = LENGTH) -> str:
    """One name, with nothing checked about it."""
    return "".join(secrets.choice(ALPHABET) for _ in range(length))


def unique_name(
    taken: Callable[[str], bool], *, length: int = LENGTH, attempts: int = ATTEMPTS
) -> str | None:
    """A name `taken` says nothing already answers to, or None if there was none.

    The check is the caller's, because what "already taken" means differs: a
    client is taken by a peer in the server config or by a file left behind in
    the clients directory, and a token by a row on the account issuing it.

    None rather than an exception, so each caller can raise the error its own
    layer knows how to word.
    """
    for _ in range(attempts):
        candidate = random_name(length)
        if not taken(candidate):
            return candidate
    return None

File: panel/frontend/src/lib/names.ts
/*
 * The random name a form starts with, when the thing being created needs one.
 *
 * A client and an API token are both named by whoever adds them, and in both
 * cases the name is an identifier: the client's is what its config file is
 * called and what every per-client route is addressed by, the token's is what
 * the activity log writes beside everything it does. An empty box in front of
 * somebody adding the fortieth client of an afternoon is a demand to invent a
 * fortieth word, and what it produces is "test2".
 *
 * So the box opens with a name already in it, and this is where it comes from.
 * The operator saves it as it stands or types over it - both are one action,
 * which is the whole point of putting a value there rather than a placeholder.
 *
 * The same nine characters and the same alphabet as the server's own generator
 * in awg/names.py, deliberately: a name suggested here and a name drawn by the
 * server when the box is left empty must be the same kind of thing, or the two
 * halves of one feature would read as two features. That file explains the
 * alphabet - lower case, and without the four characters that get read wrong
 * when a name is read off a screen and typed back somewhere else.
 *
 * This is not the collision check. Nothing in a browser can make one: the list
 * is paginated, it is seconds old, and another admin may be adding a client in
 * the next room. The server checks, under its config lock, and refuses a name
 * that is taken - which for a name drawn from forty-five bits is a reply nobody
 * will ever see, and is still the reply that makes the name an identifier.
 */

const ALPHABET = "abcdefghijkmnpqrstuvwxyz23456789";

export const NAME_LENGTH = 9;

/**
 * One suggestion.
 *
 * From `crypto.getRandomValues` rather than `Math.random`, for the reason the
 * server uses `secrets`: nothing here is a secret, but a suggestion that
 * repeats - across two tabs opened at the same moment, or a browser that seeds
 * a weak generator predictably - is exactly the collision the name is supposed
 * not to have. The modulo is unbiased because the alphabet's length divides 256.
 */
export function randomName(length: number = NAME_LENGTH): string {
  const bytes = new Uint8Array(length);
  crypto.getRandomValues(bytes);
  return Array.from(bytes, (byte) => ALPHABET[byte % ALPHABET.length]).join("");
}

/**
 * Whether a key event is somebody typing a character into the box.
 *
 * Which is what a suggestion should give way to. Not an arrow key, not a
 * backspace - those are somebody editing what is there, and the length of
 * `key` is what tells the two apart - and not a shortcut, where the character
 * belongs to the browser rather than to the field.
 *
 * Structurally typed rather than taking React's event, so that this file goes
 * on knowing nothing about the framework.
 */
export function typedCharacter(event: {
  key: string;
  ctrlKey: boolean;
  metaKey: boolean;
  altKey: boolean;
}): boolean {
  return event.key.length === 1 && !event.ctrlKey && !event.metaKey && !event.altKey;
}

/**
 * Have the character about to arrive replace the whole suggestion, rather than
 * land next to it.
 *
 * Called from the box's own keydown and paste, a moment before the browser
 * inserts anything - selecting the text there and then is what makes the
 * insertion replace it, and nothing is on screen long enough to be seen.
 *
 * The obvious way to get the same effect is to select the box when the dialog
 * opens, and it is worse: nine characters nobody typed, sitting highlighted in
 * a form that has only just appeared, read as something the panel is warning
 * about rather than as a value offered. So the box opens quietly and the
 * replacement waits for the keystroke that needs it.
 *
 * Only while the box still holds the suggestion exactly as it was drawn, and
 * only with the caret collapsed: a value somebody has edited is theirs, and a
 * selection they made by hand is a decision about which part of it to change.
 */
export function replaceSuggestion(box: HTMLInputElement, suggestion: string): void {
  if (suggestion && box.value === suggestion && box.selectionStart === box.selectionEnd) {
    box.select();
  }
}
